pageserver/src/walredo.rs

```diff
@@ -587,419 +583,1506 @@ impl<C: CommandExt> CloseFileDescriptors for C {
+                // we could send it back to the queue at least here, but not so easily when it's
+                // inflight.
+
+                if let Err(e) =
+                    response.send(Err(io_or_timeout).context("Failed to write request to wal-redo"))
+                {
+                    let e = e.into_inner().expect_err("just created the value");
+                    // log this here, even though we could return an error
+                    warn!("stopping due to io or timeout error: {e:#}");
                 }
+                // we can still continue processing pipelined requests, if any. the
+                // stdout task will exit upon seeing we've dropped the result_txs.
+                return Ok(());
             }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum StdoutTaskError {
+    ReadFailed(std::io::Error),
+    StdoutClosed,
+    ReadTimeout,
+}
 
-            // If we have some data in stdout, read it to the result buffer.
-            let out_revents = pollfds[0].revents().unwrap();
-            if out_revents & (PollFlags::POLLERR | PollFlags::POLLIN) != PollFlags::empty() {
-                nresult += self.stdout.read(&mut resultbuf[nresult..])?;
-            } else if out_revents.contains(PollFlags::POLLHUP) {
-                return Err(Error::new(
-                    ErrorKind::BrokenPipe,
-                    "WAL redo process closed its stdout unexpectedly",
-                ));
+/// Always eventually fails, returning a developer friendly reason.
+#[instrument(name = "stdout", skip_all)]
+async fn stdout_task<AR>(
+    mut stdout: AR,
+    mut inter_rx: tokio::sync::mpsc::Receiver<IntermediatePayload>,
+) -> Result<(), StdoutTaskError>
+where
+    AR: tokio::io::AsyncRead + Unpin,
+{
+    use tokio::io::AsyncReadExt;
+
+    async fn read_page<AR>(mut stdout: AR, buffer: &mut Vec<u8>) -> Result<(), StdoutTaskError>
+    where
+        AR: tokio::io::AsyncRead + Unpin,
+    {
+        loop {
+            let read = stdout
+                .read_buf(buffer)
+                .await
+                .map_err(StdoutTaskError::ReadFailed)?;
+            if read == 0 {
+                return Err(StdoutTaskError::StdoutClosed);
             }
+            if buffer.len() < 8192 {
+                continue;
+            }
+            return Ok(());
         }
+    }
 
-        Ok(Bytes::from(resultbuf))
+    // the extra past one block is for normally reading and clearing out the readyness for the
+    // stdout. tokio does not expose an api on ChildStdout for this, so we need to have this
+    // "hack". this buffer is reused between reads.
+    let mut page_buf = Vec::with_capacity(8192 + 4096);
+
+    loop {
+        let (completion, timeout_at) = match inter_rx.recv().await {
+            Some(t) => t,
+            None => return Ok(()),
+        };
+
+        let res = tokio::time::timeout_at(timeout_at, read_page(&mut stdout, &mut page_buf))
```

Comment:
> `BytesMut`

I must've meant `BufMut`.

> let's assert that it's len() is 0.

I don't think that would be a good idea. There's a test case which shows that this read works fine, even if reading the 3*4096 at once then the last 4096.

> line before this that makes it obvious we're using `page_buf` as `BytesMut`?

You mean a comment? I don't think it's too relevant, but yeah why not. The method used only accepts BufMut though, so, unsure. Also, I don't want to make this a `BytesMut` because that would then create a reference counted 3*4096 long `Bytes` out of first two consecutive pages, until degrading to some other buffering which I haven't tested.

I was hoping the the `page_buf.drain(..8192)` would be explicit enough.